scvi/module/_vae.py

```diff
@@ -289,6 +300,158 @@ def generative(
+        Gaussian kernel with :math:`\gamma`=1.
+
+        Parameters
+        ----------
+        z1
+            First tensor
+        z2
+            Second tensor
+
+        Returns
+        -------
+        Tensor containing the kernels of ``z1`` and ``z2``.
+        """
+        z1_size = z1.size(0)
+        z2_size = z2.size(0)
+
+        d = z1.size(1)
+        if d != z2.size(1):
+            raise ValueError(
+                "z1 and z2 must be defined on the same space, "
+                "but input was: z1_d={} while "
+                "z2_d={}.".format(z1.size(1), z2.size(1))
+            )
+
+        z1 = z1.unsqueeze(1)  # (z1_size, 1, d)
+        z2 = z2.unsqueeze(0)  # (1, z2_size, d)
+        z1 = z1.expand(z1_size, z2_size, d)
+        z2 = z2.expand(z1_size, z2_size, d)
+
+        exp_term = (z1 - z2).pow(2).sum(dim=2)  # (z1_size, z2_size)
+        return torch.exp(-exp_term)
+
+    def _compute_mmd(self, z1: torch.Tensor, z2: torch.Tensor) -> torch.Tensor:
+        """
+        Computes the Maximum Mean Discrepancy (MMD) of ``z1`` and ``z2`` as described in [Gretton12]_.
+        Based on https://github.com/napsternxg/pytorch-practice/blob/master/Pytorch%20-%20MMD%20VAE.ipynb
+
+        Parameters
+        ----------
+        z1
+            First tensor
+        z2
+            Second tensor
+
+        Returns
+        -------
+        Tensor with one item containing the MMD of ``z1`` and ``z2``.
+        """
+        k_z1_z1 = self._compute_mmd_kernels(z1, z1)
+        k_z2_z2 = self._compute_mmd_kernels(z2, z2)
+        k_z1_z2 = self._compute_mmd_kernels(z1, z2)
+        mmd = k_z1_z1.mean() + k_z2_z2.mean() - 2 * k_z1_z2.mean()
+        return mmd
+
+    def _compute_fast_mmd(self, z1: torch.Tensor, z2: torch.Tensor) -> torch.Tensor:
+        """
+        Computes a fast approximation of the MMD. See `_compute_mmd`.
+
+        Parameters
+        ----------
+        z1
+            First tensor
+        z2
+            Second tensor
+
+        Returns
+        -------
+        Tensor with one item containing an approximation of the MMD of ``z1`` and ``z2``.
+        """
+        z1_size = z1.size(0)
+        z2_size = z2.size(0)
+
+        # z1_size and z2_size must match, otherwise pick their min
+        batch_size = min(z1_size, z2_size)
+
+        # Drop a sample if batch_size is not even
+        if batch_size % 2 != 0:
+            batch_size -= 1
+
+        z1 = z1[:batch_size, :]
+        z2 = z2[:batch_size, :]
+
+        if z1.size(1) != z2.size(1):
+            raise ValueError(
+                "z1 and z2 must be defined on the same space, "
+                "but input was: z1_d={} while "
+                "z2_d={}.".format(z1.size(1), z2.size(1))
+            )
+
+        # zi_even contains every other sample in zi starting at 0
+        z1_even = z1[: batch_size - 1 : 2, :]
+        z1_odd = z1[1:batch_size:2, :]
+        z2_even = z2[: batch_size - 1 : 2, :]
+        z2_odd = z2[1:batch_size:2, :]
```

Comment:
no need to take every other, as the order in this tensor is random anyway, so you can like take the first half and the second half

Reply:
Thank you. I made that change. Although, I’m curious: If the order is random, shouldn’t it anyway not matter which pairing scheme we use, whether it is every other or pairs from first/second halves?